src/main/kotlin/edu/kit/compiler/semantic/visitor/NameAnalysis.kt

```diff
@@ -0,0 +1,481 @@
+        currentClass = classDeclaration
+
+        super.visitClassDeclaration(classDeclaration)
+    }
+
+    override fun visitFieldDeclaration(fieldDeclaration: AstNode.ClassMember.FieldDeclaration) {
+        val type = fieldDeclaration.type
+        if (type is SemanticType.Class) {
+            type.definition = lookupClass(global, sourceFile, type)
+        }
+    }
+
+    private inline fun <reified T : AstNode.ClassMember.SubroutineDeclaration> handleMethod(definition: T, isStatic: Boolean) {
+        val namespace = NameResolutionHelper(currentClass, sourceFile, systemSymbol, isStatic)
+        namespace.registerParameters(definition.parameters)
+        definition.accept(SubroutineNameResolver(namespace, sourceFile))
+        namespace.unregisterParameters()
+    }
+
+    override fun visitMethodDeclaration(methodDeclaration: AstNode.ClassMember.SubroutineDeclaration.MethodDeclaration) {
+        handleMethod(methodDeclaration, false)
+    }
+
+    override fun visitMainMethodDeclaration(mainMethodDeclaration: AstNode.ClassMember.SubroutineDeclaration.MainMethodDeclaration) {
+        handleMethod(mainMethodDeclaration, true)
+    }
+}
+
+class SubroutineNameResolver(
+    private val namespace: NameResolutionHelper,
+    private val sourceFile: AnnotatableFile
+) : AbstractVisitor() {
+
+    override fun visitBlock(block: AstNode.Statement.Block) {
+        namespace.enterScope()
+        super.visitBlock(block)
+        namespace.leaveScope()
+    }
+
+    override fun visitLocalVariableDeclaration(localVariableDeclaration: AstNode.Statement.LocalVariableDeclaration) {
+        namespace.addDefinition(localVariableDeclaration)
+        super.visitLocalVariableDeclaration(localVariableDeclaration)
+    }
+
+    override fun visitIdentifierExpression(identifierExpression: AstNode.Expression.IdentifierExpression) {
+        identifierExpression.definition = namespace.lookupVariable(
+            identifierExpression.name
+        )
+    }
+
+    override fun visitFieldAccessExpression(fieldAccessExpression: AstNode.Expression.FieldAccessExpression) {
+        super.visitFieldAccessExpression(fieldAccessExpression)
+
+        fieldAccessExpression.definition = namespace.lookupField(
+            fieldAccessExpression.field, fieldAccessExpression.target.actualType
+        )
+    }
+
+    override fun visitMethodInvocationExpression(methodInvocationExpression: AstNode.Expression.MethodInvocationExpression) {
+        val handled = tryHandleSystemCall(methodInvocationExpression)
+        if (handled) {
+            return
+        }
+
+        super.visitMethodInvocationExpression(methodInvocationExpression)
+
+        methodInvocationExpression.type = namespace.lookupMethod(
+            methodInvocationExpression.method, methodInvocationExpression.target?.actualType
+        )?.let { AstNode.Expression.MethodInvocationExpression.Type.Normal(it) }
+    }
+
+    companion object {
```

Comment:
Member-sortierung wäre toll :D

Reply:
Ja, kotlin braucht static felder oder iwas vergleichbares..

Reply:
naja gibt irgendeine Best-Practise Spezifikation, die besagt dass Companion Objects ganz oben in der Klasse stehen